File: nextgenie/nextgenie/page/nextgenie/nextgenie.js
// Copyright (c) 2025, Codeventurers and contributors
// For license information, please see license.txt

frappe.pages['nextgenie'].on_page_load = function (wrapper) {

	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'NextGenie',
		single_column: true
	});

	// Hide Frappe page title bar
	setTimeout(function () {
		$(wrapper).find('.page-head').hide();
		$(wrapper).closest('.page-wrapper').find('.page-head').hide();
	}, 0);

	page.main.css({ 'padding': '0', 'margin': '0' });
	$(page.main).addClass('ng-page-main');

	$(frappe.render_template("nextgenie")).appendTo(page.main);

	// ── State ────────────────────────────────────────
	var sessions = [];
	var activeId = null;

	// ── Helpers ──────────────────────────────────────

	function getTime() {
		return new Date().toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
	}

	function scrollBottom() {
		var el = document.getElementById('chat-messages');
		if (el) el.scrollTop = el.scrollHeight;
	}

	// ── Render history list ───────────────────────────

	function renderHistory() {
		var $list = $('#hist-list').empty();
		sessions.forEach(function (s) {
			var $item = $('<div>')
				.addClass('hist-item')
				.toggleClass('active', s.id === activeId);

			$('<div>').addClass('hist-item-title').text(s.title).appendTo($item);
			$('<div>').addClass('hist-item-date').text(s.date || '').appendTo($item);

			$item.on('click', function () { loadSession(s.id); });
			$list.append($item);
		});
	}

	// ── Load session messages ─────────────────────────

	function loadSession(id) {
		activeId = id;
		var s = sessions.find(function (x) { return x.id === id; });
		if (s) $('#chat-header-title').text(s.title);
		$('#chat-messages').empty();
		renderHistory();

		frappe.call({
			method: 'nextgenie.nextgenie.page.nextgenie.nextgenie.get_session_messages',
			args: { session_id: id },
			callback: function (r) {
				if (r.message && r.message.length) {
					r.message.forEach(function (m) {
						appendMessage(m.sender, m.message, false);
					});
				}
			}
		});
	}

	// ── Append message bubble ─────────────────────────

	function appendMessage(sender, text, animate) {
		if (animate === undefined) animate = true;
		var isUser = (sender === 'User');

		// Row
		var $row = $('<div>').addClass('ng-msg-row');
		if (isUser) $row.addClass('ng-user');
		if (!animate) $row.css('animation', 'none');

		// Avatar
		var $av = $('<div>')
			.addClass('ng-avatar')
			.addClass(isUser ? 'ng-av-user' : 'ng-av-bot')
			.text(isUser ? '👤' : '✦');

		// Bubble wrapper
		var $wrap = $('<div>').addClass('ng-bubble-wrap');

		// Bubble — use .text() so no HTML injection and no char-wrap
		var $bub = $('<div>')
			.addClass('ng-bubble')
			.addClass(isUser ? 'ng-usr' : 'ng-bot')
			.text(text);

		// Timestamp
		var $time = $('<div>').addClass('ng-time').text(getTime());

		$wrap.append($bub).append($time);
		$row.append($av).append($wrap);
		$('#chat-messages').append($row);
		scrollBottom();
	}

	function showTyping() {
		var $row = $('<div>').addClass('ng-msg-row').attr('id', 'ng-typing-row');
		var $av = $('<div>').addClass('ng-avatar ng-av-bot').text('✦');
		var $typ = $('<div>').addClass('ng-typing')
			.append($('<span>')).append($('<span>')).append($('<span>'));
		$row.append($av).append($typ);
		$('#chat-messages').append($row);
		scrollBottom();
	}

	function hideTyping() {
		$('#ng-typing-row').remove();
	}

	// ── Load all sessions ─────────────────────────────

	function loadSessions() {
		frappe.call({
			method: 'nextgenie.nextgenie.page.nextgenie.nextgenie.get_sessions',
			callback: function (r) {
				if (r.message && r.message.length) {
					sessions = r.message;
					renderHistory();
					loadSession(sessions[0].id);
				} else {
					createNewSession();
				}
			}
		});
	}

	// ── Create new session ────────────────────────────

	function createNewSession() {
		frappe.call({
			method: 'nextgenie.nextgenie.page.nextgenie.nextgenie.create_session',
			callback: function (r) {
				if (r.message) {
					sessions.unshift(r.message);
					activeId = r.message.id;
					$('#chat-messages').empty();
					$('#chat-header-title').text('New Chat');
					renderHistory();

					var welcome = "👋 Hello! I'm NextGenie, your private HR assistant. How can I help you today?";
					appendMessage('Genie', welcome);
					frappe.call({
						method: 'nextgenie.nextgenie.page.nextgenie.nextgenie.add_chat_message',
						args: { session_id: activeId, sender: 'Genie', message: welcome }
					});
				}
			}
		});
	}

	// ── Send message ──────────────────────────────────

	function sendMessage() {
		var $input = $('#chat-input-box');
		var message = $input.val().trim();
		if (!message) return;

		appendMessage('User', message);
		$input.val('');

		// Auto-rename on first message
		var s = sessions.find(function (x) { return x.id === activeId; });
		if (s && s.title === 'New Chat') {
			s.title = message.length > 28 ? message.slice(0, 28) + '…' : message;
			$('#chat-header-title').text(s.title);
			renderHistory();
			frappe.call({
				method: 'nextgenie.nextgenie.page.nextgenie.nextgenie.update_session_title',
				args: { session_id: activeId, title: s.title }
			});
		}

		// Save user message
		frappe.call({
			method: 'nextgenie.nextgenie.page.nextgenie.nextgenie.add_chat_message',
			args: { session_id: activeId, sender: 'User', message: message }
		});

		showTyping();

		// Get AI reply
		frappe.call({
			method: 'nextgenie.nextgenie.page.nextgenie.nextgenie.get_ai_reply',
			args: { session_id: activeId, message: message },
			callback: function (r) {
				hideTyping();
				var reply = (r.message && r.message.reply)
					? r.message.reply
					: "I'm here to help! Could you clarify your question?";
				appendMessage('Genie', reply);
				frappe.call({
					method: 'nextgenie.nextgenie.page.nextgenie.nextgenie.add_chat_message',
					args: { session_id: activeId, sender: 'Genie', message: reply }
				});
			},
			error: function () {
				hideTyping();
				appendMessage('Genie', '⚠️ Something went wrong. Please try again.');
			}
		});
	}

	// ── Events ────────────────────────────────────────

	$(document).on('click', '#send-btn', sendMessage);

	$(document).on('keypress', '#chat-input-box', function (e) {
		if (e.which === 13) sendMessage();
	});

	$(document).on('click', '#new-chat-btn', createNewSession);

	$(document).on('click', '#clear-btn', function () {
		if (confirm('Clear this chat?')) {
			frappe.call({
				method: 'nextgenie.nextgenie.page.nextgenie.nextgenie.clear_session_messages',
				args: { session_id: activeId },
				callback: function () {
					$('#chat-messages').empty();
					appendMessage('Genie', 'Chat cleared. How can I assist you?');
				}
			});
		}
	});

	$(document).on('click', '#quick-actions li', function () {
		var msg = $(this).data('msg');
		if (msg) {
			$('#chat-input-box').val(msg);
			sendMessage();
		}
	});

	// ── Boot ─────────────────────────────────────────
	loadSessions();
};

File: nextgenie/nextgenie/page/nextgenie/nextgenie.py
# Copyright (c) 2025, Codeventurers and contributors
# For license information, please see license.txt

import frappe
import requests
from datetime import datetime


# ── Session Management ────────────────────────────────────────


@frappe.whitelist()
def get_sessions():
    """Return all chat sessions for the current user, newest first."""
    sessions = frappe.get_all(
        "Genie Chat Session",
        filters={"owner": frappe.session.user},
        fields=["name as id", "title", "creation"],
        order_by="creation desc",
        limit=30
    )
    for s in sessions:
        s["date"] = _friendly_date(s["creation"])
    return sessions


@frappe.whitelist()
def create_session():
    """Create a new chat session and return it."""
    doc = frappe.get_doc({
        "doctype": "Genie Chat Session",
        "title": "New Chat",
        "owner": frappe.session.user
    })
    doc.insert(ignore_permissions=True)
    frappe.db.commit()
    return {
        "id":    doc.name,
        "title": doc.title,
        "date":  "Just now"
    }


@frappe.whitelist()
def update_session_title(session_id, title):
    """Rename a session."""
    frappe.db.set_value("Genie Chat Session", session_id, "title", title)
    frappe.db.commit()
    return {"status": "ok"}


# ── Message Management ────────────────────────────────────────


@frappe.whitelist()
def get_session_messages(session_id):
    """Return all messages for a session."""
    return frappe.get_all(
        "Genie Chat Message",
        filters={"session": session_id},
        fields=["sender", "message", "creation"],
        order_by="creation asc"
    )


@frappe.whitelist()
def add_chat_message(session_id, sender, message):
    """Save a single chat message."""
    doc = frappe.get_doc({
        "doctype": "Genie Chat Message",
        "session": session_id,
        "sender":  sender,
        "message": message
    })
    doc.insert(ignore_permissions=True)
    frappe.db.commit()
    return {"status": "ok"}


@frappe.whitelist()
def clear_session_messages(session_id):
    """Delete all messages in a session."""
    frappe.db.delete("Genie Chat Message", {"session": session_id})
    frappe.db.commit()
    return {"status": "ok"}


# ── AI Reply ──────────────────────────────────────────────────


@frappe.whitelist()
def get_ai_reply(session_id, message):
    """
    Get AI reply from local Ollama model.
    Ollama must be running: ollama serve
    Model must be pulled:   ollama pull phi3:mini
    """

    # Build conversation history for context
    history = frappe.get_all(
        "Genie Chat Message",
        filters={"session": session_id},
        fields=["sender", "message"],
        order_by="creation asc",
        limit=20  # last 20 messages as context
    )

    # Format history as prompt context
    context = ""
    for msg in history:
        role = "User" if msg["sender"] == "User" else "Assistant"
        context += f"{role}: {msg['message']}\n"
    context += f"User: {message}\nAssistant:"

    system_prompt = """You are NextGenie, a helpful and professional HR assistant.
You help employees with leave management, payroll queries, attendance,
onboarding, company policies, and HR-related questions.
Always be polite, concise, and accurate. If you don't know something,
say so and suggest contacting the HR team directly."""

    try:
        response = requests.post(
            "http://localhost:11434/api/generate",
            json={
                "model":  "phi3:mini",   # change to your pulled model
                "system": system_prompt,
                "prompt": context,
                "stream": False,
                "options": {
                    "temperature": 0.7,
                    "num_predict": 300
                }
            },
            timeout=60
        )
        data = response.json()
        reply = data.get("response", "").strip()
        if not reply:
            reply = "I'm here to help. Could you clarify your question?"
        return {"reply": reply}

    except requests.exceptions.ConnectionError:
        return {"reply": "⚠️ Local AI is offline. Please start Ollama: `ollama serve`"}
    except Exception as e:
        frappe.log_error(str(e), "NextGenie AI Error")
        return {"reply": "⚠️ Something went wrong. Please try again."}


# ── Utils ─────────────────────────────────────────────────────


def _friendly_date(creation):
    """Convert creation datetime to a friendly label."""
    if not creation:
        return ""
    # creation may be a string from frappe — parse it
    if isinstance(creation, str):
        try:
            creation = datetime.strptime(creation, "%Y-%m-%d %H:%M:%S.%f")
        except ValueError:
            creation = datetime.strptime(creation, "%Y-%m-%d %H:%M:%S")
    now   = datetime.now()
    delta = now - creation
    total_seconds = int(delta.total_seconds())
    if total_seconds < 60:
        return "Just now"
    elif total_seconds < 3600:
        mins = total_seconds // 60
        return f"{mins}m ago"
    elif total_seconds < 86400:
        hrs = total_seconds // 3600
        return f"{hrs}h ago"
    elif total_seconds < 172800:
        return "Yesterday"
    elif delta.days < 7:
        return f"{delta.days} days ago"
    else:
        return creation.strftime("%d %b %Y")